rllib/agents/alpha_star/alpha_star.py

```diff
@@ -0,0 +1,456 @@
+        replay_actor_args = [
+            self.config["replay_buffer_capacity"],
+            self.config["replay_buffer_replay_ratio"],
+        ]
+
+        # Create a DistributedLearners utility object and set it up with
+        # the initial first n learnable policies (found in the config).
+        distributed_learners = DistributedLearners(
+            config=self.config,
+            max_num_policies_to_train=self.config["max_num_policies_to_train"],
+            replay_actor_class=ReplayActor,
+            replay_actor_args=replay_actor_args,
+        )
+        for pid, policy_spec in ma_cfg["policies"].items():
+            if pid in self.workers.local_worker().get_policies_to_train():
+                distributed_learners.add_policy(pid, policy_spec)
+
+        # Store distributed_learners on all RolloutWorkers
+        # so they know, to which replay shard to send samples to.
+
+        def _set_policy_learners(worker):
+            worker._distributed_learners = distributed_learners
+
+        ray.get(
+            [
+                w.apply.remote(_set_policy_learners)
+                for w in self.workers.remote_workers()
+            ]
+        )
+
+        self.distributed_learners = distributed_learners
+
+        # Store the win rates for league overview printouts.
+        self.win_rates: DefaultDict[PolicyID, float] = defaultdict(float)
+
+    @override(Trainer)
+    def step(self) -> ResultDict:
+        # Perform a full step (including evaluation).
+        result = super().step()
+
+        # Based on the (train + evaluate) results, perform a step of
+        # league building.
+        self.league_builder.build_league(result=result)
+
+        return result
+
+    @override(Trainer)
+    def training_iteration(self) -> ResultDict:
+        # Trigger asynchronous rollouts on all RolloutWorkers.
+        # - Rollout results are sent directly to correct replay buffer
+        #   shards, instead of here (to the driver).
+        with self._timers[SAMPLE_TIMER]:
+            sample_results = asynchronous_parallel_requests(
+                remote_requests_in_flight=self.remote_requests_in_flight,
+                actors=self.workers.remote_workers() or [self.workers.local_worker()],
+                ray_wait_timeout_s=0.01,
+                max_remote_requests_in_flight_per_actor=2,
+                remote_fn=self._sample_and_send_to_buffer,
+            )
+        # Update sample counters.
+        for (env_steps, agent_steps) in sample_results.values():
+            self._counters[NUM_ENV_STEPS_SAMPLED] += env_steps
+            self._counters[NUM_AGENT_STEPS_SAMPLED] += agent_steps
+
+        # Trigger asynchronous training update requests on all learning
+        # policies.
+        with self._timers[LEARN_ON_BATCH_TIMER]:
+            pol_actors = []
+            args = []
+            for i, (pid, pol_actor, repl_actor) in enumerate(self.distributed_learners):
```

Comment:
why bother with enumerate here? doesn't seem like i is used for anything?

Reply:
fixed